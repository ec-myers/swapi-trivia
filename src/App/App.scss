$accent-color-dark: rgba(108, 91, 117, 0.8);
$accent-color-light: rgba(242,136,79, 0.8);
@import url('https://fonts.googleapis.com/css?family=IBM+Plex+Mono:300,400,500,700&display=swap');

* {
  box-sizing: border-box;
  font-family: 'IBM Plex Mono', monospace;
  margin: 0;
  padding: 0
}

.App {
  background: url(../Images/tatooine_background.jpg); 
  background-size: cover;
  background-repeat: no-repeat
}

