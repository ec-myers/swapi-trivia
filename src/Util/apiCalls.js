export const getFilms = () => {
  return fetch('https://swapi.co/api/films')
    .then(response => response.json())
    .then(data => sortFilmsByReleaseDate(data))
    .catch(error => console.log(error));
}

const sortFilmsByReleaseDate = (data) => {
  return data.results.sort((a, b) => {
    let yearA = parseInt(a.release_date.slice(0, 4));
    let yearB = parseInt(b.release_date.slice(0, 4));
    return yearA - yearB
  }).map(movie => {
    return {
      title: movie.title,
      episode: movie.episode_id,
      releaseYear: movie.release_date.slice(0, 4)
    }
  })
}