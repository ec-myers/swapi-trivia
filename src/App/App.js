import React, { Component } from 'react';
import { getFilms } from '../Util/apiCalls';
import { Link, BrowserRouter, Switch, Route } from 'react-router-dom';
import '../App/App.scss';
import Form from '../Form/Form';
import Container from '../Container/Container';
import Nav from '../Nav/Nav';


class App extends Component {
  constructor() {
    super();
    this.state = {
      userInfo: {},
      movies:[],
      isFormComplete: false
    }
  }

  componentDidMount() {
    getFilms().then(data => this.setState({movies: data}))
  }

  getFormData = (userInfo) => {
    this.setState({userInfo: userInfo, isFormComplete:true})
  }

  render() {
    const{movies, isFormComplete, userInfo} = this.state
    return (
      <>
        {!isFormComplete && <Form getFormData={this.getFormData}/>}
        {isFormComplete && <Nav user={userInfo}/>}
        {isFormComplete && <Container movies={movies}/>}
      </>
    )
  }
}

export default App;
