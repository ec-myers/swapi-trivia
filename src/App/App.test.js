import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { getFilms } from '../Util/apiCalls';

jest.mock('../Util/apiCalls.js');

describe('App', () => {
  let wrapper, 
      mockFilms;
  
  beforeEach(() => {
    mockFilms = [
      {
        title: 'A New Hope',
        scrollText: 'blah blah',
        id: 1,
        releaseYear: '1974',
        episode: 4
      }, {
        title: 'Whatever',
        scrollText: 'blah blah',
        id: 3,
        releaseYear: '1985',
        episode: 7
      }
    ]
  
    getFilms.mockImplementation(() => {
      return Promise.resolve(mockFilms);
    });
    wrapper = shallow(<App />);
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should retrieve films after mounting', () => {
    shallow(<App />);
    expect(getFilms).toHaveBeenCalled();
  });

  it('should update state with characters when called', () => {
    let user = {
      name: 'Greg',
      quote: 'Yolo',
      rank: 'Sucks'
    };

    window.localStorage = jest.fn();

    expect(wrapper.state('userInfo')).toEqual({});
    expect(wrapper.state('isFormComplete')).toEqual(false);
    wrapper.instance().getFormData(user);
    expect(wrapper.state('userInfo')).toEqual(user);
    expect(wrapper.state('isFormComplete')).toEqual(true);
  });

  it('should update state when new character is favorited', () => {
    let character = {
      name: 'Steve',
      homeworld: 'Dantooine',
      population: 200000,
      species: "Twi'lek",
      films: ['movie', 'other movie']
    };

    expect(wrapper.state('favorites')).toEqual([]);
    wrapper.instance().toggleFavorite(character);
    expect(wrapper.state('favorites')).toEqual([character])
  });

  it('should update state when existing character is favorited ', () => {
    let character = {
      name: 'Steve',
      homeworld: 'Dantooine',
      population: 200000,
      species: "Twi'lek",
      films: ['movie', 'other movie']
    };

    wrapper.instance().setState({favorites: [character]});
    expect(wrapper.state('favorites')).toEqual([character]);
    wrapper.instance().toggleFavorite(character);
    expect(wrapper.state('favorites')).toEqual([])
  });

  it('should update state when user logs out', () => {
    let user = {
      name: 'Greg',
      quote: 'Yolo',
      rank: 'Sucks'
    };

    wrapper.instance().setState({userInfo: user});
    expect(wrapper.state('userInfo')).toEqual(user);
    expect(wrapper.state('isFormComplete')).toEqual(true);
    wrapper.instance().logOut();
    expect(wrapper.state('userInfo')).toEqual({});
    expect(wrapper.state('isFormComplete')).toEqual(false);
  });


});