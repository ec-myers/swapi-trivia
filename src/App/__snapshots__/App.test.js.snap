// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<main
  className="App"
>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
</main>
`;

exports[`App with no movies should render the loading img if no movies are present 1`] = `
<main
  className="App"
>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <div
    className="loading-img"
  />
  <Route
    exact={true}
    path="/movies"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/movies/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
</main>
`;
